File: app/src/main/java/com/wa2c/android/medoly/plugin/action/viewlyrics/activity/SearchActivity.java
package com.wa2c.android.medoly.plugin.action.viewlyrics.activity;

import android.app.ActionBar;
import android.app.Activity;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ListView;
import android.widget.RadioButton;
import android.widget.ScrollView;
import android.widget.TextView;

import com.cybozu.labs.langdetect.Detector;
import com.cybozu.labs.langdetect.Language;
import com.wa2c.android.medoly.plugin.action.viewlyrics.R;
import com.wa2c.android.medoly.plugin.action.viewlyrics.db.SearchCacheHelper;
import com.wa2c.android.medoly.plugin.action.viewlyrics.dialog.ConfirmDialogFragment;
import com.wa2c.android.medoly.plugin.action.viewlyrics.search.Result;
import com.wa2c.android.medoly.plugin.action.viewlyrics.search.ResultItem;
import com.wa2c.android.medoly.plugin.action.viewlyrics.search.ViewLyricsSearcher;
import com.wa2c.android.medoly.plugin.action.viewlyrics.util.AppUtils;
import com.wa2c.android.medoly.plugin.action.viewlyrics.util.Logger;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.Background;
import org.androidannotations.annotations.Click;
import org.androidannotations.annotations.EActivity;
import org.androidannotations.annotations.Extra;
import org.androidannotations.annotations.ItemClick;
import org.androidannotations.annotations.OptionsItem;
import org.androidannotations.annotations.OptionsMenu;
import org.androidannotations.annotations.UiThread;
import org.androidannotations.annotations.ViewById;

import java.io.BufferedWriter;
import java.io.IOException;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Locale;

/**
 * Search Activity.
 */
@EActivity(R.layout.activity_search)
@OptionsMenu(R.menu.activity_search)
public class SearchActivity extends Activity {

    private static final int REQUEST_CODE_SAVE_FILE = 1;
    public static final String INTENT_SEARCH_TITLE = "INTENT_SEARCH_TITLE";
    public static final String INTENT_SEARCH_ARTIST = "INTENT_SEARCH_ARTIST";

    /** Search list adapter. */
    private SearchResultAdapter searchResultAdapter;
    /** Search cache helper. */
    private SearchCacheHelper searchCacheHelper;

    @Extra(INTENT_SEARCH_TITLE)
    String intentSearchTitle;
    @Extra(INTENT_SEARCH_ARTIST)
    String intentSearchArtist;

    @ViewById
    Button searchTitleButton;
    @ViewById
    EditText searchTitleEditText;
    @ViewById
    Button searchArtistButton;
    @ViewById
    EditText searchArtistEditText;
    @ViewById
    ImageButton searchClearButton;
    @ViewById
    ImageButton searchStartButton;
    @ViewById
    ListView searchResultListView;
    @ViewById
    ScrollView searchLyricsScrollView;
    @ViewById
    TextView searchLyricsTextView;

    @ViewById
    View searchResultLoadingLayout;
    @ViewById
    View searchLyricsLoadingLayout;

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        setIntent(intent);
    }

    @AfterViews
    void afterViews() {
        ActionBar actionBar = getActionBar();
        if (actionBar != null) {
            actionBar.setDisplayShowHomeEnabled(true);
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowTitleEnabled(true);
        }

        searchCacheHelper = new SearchCacheHelper(this);
        searchResultAdapter = new SearchResultAdapter(this);
        searchResultListView.setAdapter(searchResultAdapter);

        searchTitleEditText.setText(intentSearchTitle);
        searchArtistEditText.setText(intentSearchArtist);
    }

    @OptionsItem(android.R.id.home)
    void menuHomeClick() {
        startActivity(new Intent(this, MainActivity_.class));
    }

    @OptionsItem(R.id.menu_search_save_file)
    void menuSaveFileClick() {
        if (!existsLyrics()) {
            AppUtils.showToast(this, R.string.error_exists_lyrics);
            return;
        }

        Intent intent = new Intent(Intent.ACTION_CREATE_DOCUMENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        intent.setType("*/*");
        intent.putExtra(Intent.EXTRA_TITLE,  searchResultAdapter.getSelectedItem().getMusicTitle() + ".lrc");
        startActivityForResult(intent, REQUEST_CODE_SAVE_FILE);
    }

    @OptionsItem(R.id.menu_search_save_cache)
    void menuSaveCacheClick() {
        if (!existsLyrics()) {
            AppUtils.showToast(this, R.string.error_exists_lyrics);
            return;
        }

        ConfirmDialogFragment dialog = ConfirmDialogFragment.newInstance(
                getString(R.string.dialog_confirm_message_save_cache),
                getString(R.string.label_confirmation),
                getString(R.string.dialog_confirm_label_save_cache),
                null,
                getString(android.R.string.cancel)
        );
        dialog.setClickListener(new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                if (which == DialogInterface.BUTTON_POSITIVE) {
                    String title = searchTitleEditText.getText().toString();
                    String artist = searchArtistEditText.getText().toString();
                    saveBackground(title, artist, searchResultAdapter.getSelectedItem());
                }
            }
        });
        dialog.show(this);
    }

    @OptionsItem(R.id.menu_search_open_cache)
    void menuOpenCacheClick() {
        Intent intent = new Intent(this, CacheActivity_.class);
        intent.putExtra(CacheActivity.INTENT_SEARCH_TITLE, searchTitleEditText.getText().toString());
        intent.putExtra(CacheActivity.INTENT_SEARCH_ARTIST, searchArtistEditText.getText().toString());
        startActivity(intent);
    }

    @Background
    void saveBackground(String title, String artist, ResultItem item) {
        if (searchCacheHelper.insertOrUpdate(title, artist, null, item))
            AppUtils.showToast(this, R.string.message_save_cache);
    }

    @Click(R.id.searchTitleButton)
    void searchTitleButtonClick(View view) {

    }

    @Click(R.id.searchArtistButton)
    void searchArtistButtonClick(View view) {

    }

    @Click(R.id.searchClearButton)
    void searchClearButtonClick() {
        searchTitleEditText.setText(null);
        searchArtistEditText.setText(null);
    }

    @Click(R.id.searchStartButton)
    void searchStartButtonClick(View view) {
        // Hide keyboard
        InputMethodManager inputMethodMgr = (InputMethodManager)getSystemService(INPUT_METHOD_SERVICE);
        inputMethodMgr.hideSoftInputFromWindow(view.getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS);

        String title = searchTitleEditText.getText().toString();
        String artist = searchArtistEditText.getText().toString();
        if (TextUtils.isEmpty(title) && TextUtils.isEmpty(artist)) {
            AppUtils.showToast(this, R.string.error_input_condition);
            return;
        }

        // Clear view
        showSearchResult(null);
        showLyrics(null);
        searchResultAdapter.setSelectedItem(null);

        searchResultListView.setVisibility(View.INVISIBLE);
        searchResultLoadingLayout.setVisibility(View.VISIBLE);
        searchLyrics(title, artist);
    }

    // Search

    @Background
    void searchLyrics(String title, String artist) {
        Result result = null;
        try {
            result = ViewLyricsSearcher.search(title, artist, 0);
        } catch (Exception e) {
            Logger.e(e);
        } finally {
            showSearchResult(result);
        }
    }

    @UiThread
    void showSearchResult(Result result) {
        try {
            searchResultAdapter.clear();
            if (result != null)
                searchResultAdapter.addAll(result.getInfoList());
            searchResultAdapter.notifyDataSetChanged();
        } finally {
            searchResultListView.setVisibility(View.VISIBLE);
            searchResultLoadingLayout.setVisibility(View.INVISIBLE);
        }
    }

    // Download

    @ItemClick(R.id.searchResultListView)
    void searchResultListViewItemClick(@NonNull ResultItem item) {
        // Hide keyboard
        InputMethodManager inputMethodMgr = (InputMethodManager)getSystemService(INPUT_METHOD_SERVICE);
        inputMethodMgr.hideSoftInputFromWindow(searchResultListView.getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS);

        // Clear view
        showLyrics(null);

        searchLyricsScrollView.setVisibility(View.INVISIBLE);
        searchLyricsLoadingLayout.setVisibility(View.VISIBLE);
        downloadLyrics(item);
    }

    @Background
    void downloadLyrics(ResultItem item) {
        try {
            if (item != null) {
                String lyrics = ViewLyricsSearcher.downloadLyricsText(item.getLyricURL());
                item.setLyrics(lyrics);

                Detector d = AppUtils.getDetector(SearchActivity.this, "ja");
                if (d != null) {
                    d.append(lyrics);
                    ArrayList<Language> list = d.getProbabilities();
                    Logger.d(list);
                }

            }
        } catch (Exception e) {
            Logger.e(e);
        } finally {
            showLyrics(item);
        }
    }

    @UiThread
    void showLyrics(ResultItem item) {
        if (item == null) {
            searchLyricsTextView.setText(null);
        } else {
            searchLyricsTextView.setText(item.getLyrics());
        }
        searchResultAdapter.setSelectedItem(item);
        searchResultAdapter.notifyDataSetChanged();
        searchLyricsScrollView.setVisibility(View.VISIBLE);
        searchLyricsLoadingLayout.setVisibility(View.INVISIBLE);
    }


    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent resultData) {
        if (requestCode == REQUEST_CODE_SAVE_FILE) {
            // 歌詞のファイル保存
            if (resultCode == RESULT_OK) {
                if (!existsLyrics()) {
                    AppUtils.showToast(this, R.string.error_exists_lyrics);
                    return;
                }

                Uri uri = resultData.getData();
                try (OutputStream stream = getContentResolver().openOutputStream(uri);
                     BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(stream,  "UTF-8"))) {
                    writer.write(searchResultAdapter.getSelectedItem().getLyrics());
                    writer.flush();
                    AppUtils.showToast(this, R.string.message_lyrics_save_succeeded);
                } catch (IOException e) {
                    Logger.e(e);
                    AppUtils.showToast(this, R.string.message_lyrics_save_failed);
                }
            }
        }

        // Hide keyboard
        if (getCurrentFocus() != null) {
            InputMethodManager inputMethodMgr = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
            inputMethodMgr.hideSoftInputFromWindow(getCurrentFocus().getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS);
        }    }

    /**
     * Check existence of lyrics
     * @return true if exists lyrics.
     */
    private synchronized boolean existsLyrics() {
        return (searchResultAdapter.getSelectedItem() != null) && !TextUtils.isEmpty(searchResultAdapter.getSelectedItem().getLyricURL());
    }

    /**
     * Search result adapter.
     */
    private static class SearchResultAdapter extends ArrayAdapter<ResultItem> {

        /** Selected item. */
        private ResultItem selectedItem = null;
        /** Set selected item. */
        void setSelectedItem(ResultItem item) {
            this.selectedItem = item;
        }
        /** Get selected item. */
        ResultItem getSelectedItem() {
            return this.selectedItem;
        }


        SearchResultAdapter(Context context) {
            super(context, R.layout.layout_search_item);
        }

        @NonNull
        @Override
        public View getView(final int position, View convertView, @NonNull final ViewGroup parent) {
            // view
            final ListView listView = (ListView)parent;
            final ListItemViewHolder holder;
            if (convertView == null) {
                final View view = View.inflate(parent.getContext(), R.layout.layout_search_item, null);
                holder = new ListItemViewHolder();
                holder.searchItemRadioButton = (RadioButton)view.findViewById(R.id.searchItemRadioButton);
                holder.searchItemTitleTextView  = (TextView)view.findViewById(R.id.searchItemTitleTextView);
                holder.searchItemArtistTextView = (TextView)view.findViewById(R.id.searchItemArtistTextView);
                holder.searchItemAlbumTextView  = (TextView)view.findViewById(R.id.searchItemAlbumTextView);
                holder.searchItemDownloadTextView = (TextView)view.findViewById(R.id.searchItemDownloadTextView);
                holder.searchItemRatingTextView = (TextView)view.findViewById(R.id.searchItemRatingTextView);
                holder.searchItemFromTextView   = (TextView)view.findViewById(R.id.searchItemFromTextView);
                view.setTag(holder);
                convertView = view;
            } else {
                holder = (ListItemViewHolder) convertView.getTag();
            }

            // data
            ResultItem item = getItem(position);
            if (item == null)
                item = new ResultItem();
            holder.searchItemRadioButton.setChecked((item == selectedItem));
            holder.searchItemTitleTextView.setText(item.getMusicTitle());
            holder.searchItemArtistTextView.setText(AppUtils.coalesce(item.getMusicArtist(), "-"));
            holder.searchItemAlbumTextView.setText(AppUtils.coalesce(item.getMusicAlbum(), "-"));
            holder.searchItemDownloadTextView.setText(getContext().getString(R.string.label_search_item_download, item.getLyricDownloadsCount()));
            holder.searchItemRatingTextView.setText(getContext().getString(R.string.label_search_item_rating, item.getLyricRate(), item.getLyricRatesCount()));
            holder.searchItemFromTextView.setText(getContext().getString(R.string.label_search_item_from, item.getLyricUploader()));

            holder.searchItemRadioButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    listView.performItemClick(v, position, getItemId(position));
                }
            });

            return convertView;
        }

    }

    /** リスト項目のビュー情報を保持するHolder。 */
    private static class ListItemViewHolder {
        RadioButton searchItemRadioButton;
        TextView searchItemTitleTextView;
        TextView searchItemArtistTextView;
        TextView searchItemAlbumTextView;
        TextView searchItemDownloadTextView;
        TextView searchItemRatingTextView;
        TextView searchItemFromTextView;
    }

}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lrclyrics/activity/SearchActivity.kt
package com.wa2c.android.medoly.plugin.action.lrclyrics.activity

import android.annotation.SuppressLint
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.*
import android.view.inputmethod.InputMethodManager
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.wa2c.android.medoly.plugin.action.lrclyrics.R
import com.wa2c.android.medoly.plugin.action.lrclyrics.databinding.ActivitySearchBinding
import com.wa2c.android.medoly.plugin.action.lrclyrics.databinding.LayoutSearchItemBinding
import com.wa2c.android.medoly.plugin.action.lrclyrics.db.AppDatabase
import com.wa2c.android.medoly.plugin.action.lrclyrics.db.SearchCache
import com.wa2c.android.medoly.plugin.action.lrclyrics.db.SearchCacheDao
import com.wa2c.android.medoly.plugin.action.lrclyrics.dialog.ConfirmDialogFragment
import com.wa2c.android.medoly.plugin.action.lrclyrics.dialog.NormalizeDialogFragment
import com.wa2c.android.medoly.plugin.action.lrclyrics.search.Result
import com.wa2c.android.medoly.plugin.action.lrclyrics.search.ResultItem
import com.wa2c.android.medoly.plugin.action.lrclyrics.search.ViewLyricsSearcher
import com.wa2c.android.medoly.plugin.action.lrclyrics.util.AppUtils
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import timber.log.Timber


/**
 * Search Activity.
 */
class SearchActivity : AppCompatActivity() {

    /** Binding. */
    private lateinit var binding: ActivitySearchBinding

    /** Search list adapter.  */
    private lateinit var searchResultAdapter: SearchResultAdapter
    /** Search cache DAO. */
    private lateinit var dao: SearchCacheDao

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_search)

        val intentSearchTitle : String? = intent.getStringExtra(INTENT_SEARCH_TITLE)
        val intentSearchArtist : String? = intent.getStringExtra(INTENT_SEARCH_ARTIST)

        binding.searchTitleEditText.setText(intentSearchTitle)
        binding.searchArtistEditText.setText(intentSearchArtist)

        // Action Bar
        supportActionBar?.let {
            it.setDisplayShowHomeEnabled(true)
            it.setDisplayHomeAsUpEnabled(true)
            it.setDisplayShowTitleEnabled(true)
        }

        dao = AppDatabase.buildDb(this).getSearchCacheDao()
        searchResultAdapter = SearchResultAdapter()
        binding.searchResultListView.adapter = searchResultAdapter
        binding.searchResultListView.layoutManager = LinearLayoutManager(this)

        val searchResultHeight = resources.getDimensionPixelSize(R.dimen.search_result_height)

        // adjust size
        binding.searchLyricsScrollView.post(Runnable {
            // adjust height
            val heightResult = binding.searchResultListView.measuredHeight
            val heightLyrics = binding.searchLyricsScrollView.measuredHeight
            val heightSum = heightResult + heightLyrics
            if (heightResult == 0)
                return@Runnable
            if (heightSum < searchResultHeight * 2) {
                val params = binding.searchResultListView.layoutParams
                params.height = heightSum / 2
                binding.searchResultListView.layoutParams = params
            }
        })

        // Title button
        binding.searchTitleButton.setOnClickListener {
            val dialogFragment = NormalizeDialogFragment.newInstance(binding.searchTitleEditText.text.toString(), intentSearchTitle)
            dialogFragment.clickListener = { _, which, bundle ->
                if (which == DialogInterface.BUTTON_POSITIVE) {
                    binding.searchTitleEditText.setText(bundle?.getString(NormalizeDialogFragment.RESULT_INPUT_TEXT))
                }
            }
            dialogFragment.show(this)
        }

        // Artist button
        binding.searchArtistButton.setOnClickListener {
            val dialogFragment = NormalizeDialogFragment.newInstance(binding.searchArtistEditText.text.toString(), intentSearchArtist)
            dialogFragment.clickListener = { _, which, bundle ->
                if (which == DialogInterface.BUTTON_POSITIVE) {
                    binding.searchArtistEditText.setText(bundle?.getString(NormalizeDialogFragment.RESULT_INPUT_TEXT))
                }
            }
            dialogFragment.show(this)

        }

        // Clear[x] button
        binding.searchClearButton.setOnClickListener {
            binding.searchTitleEditText.text = null
            binding.searchArtistEditText.text = null
        }

        // Search button
        binding.searchStartButton.setOnClickListener { view ->
            // Hide keyboard
            val inputMethodMgr = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            inputMethodMgr.hideSoftInputFromWindow(view.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)

            val title = binding.searchTitleEditText.text.toString()
            val artist = binding.searchArtistEditText.text.toString()
            if (title.isEmpty() && artist.isEmpty()) {
                AppUtils.showToast(this, R.string.error_input_condition)
                return@setOnClickListener
            }

            // Clear view
            showSearchResult(null)
            searchResultAdapter.selectedItem = null

            binding.searchResultListView.visibility = View.INVISIBLE
            binding.searchResultLoadingLayout.visibility = View.VISIBLE

            //searchLyrics(title, artist)
            GlobalScope.launch(Dispatchers.Main) {
                val result = async(Dispatchers.Default) {
                    try {
                        return@async ViewLyricsSearcher.search(title, artist, 0)
                    } catch (e: Exception) {
                        return@async null
                    }
                }
                val r = result.await()
                if (r == null)
                    AppUtils.showToast(applicationContext, R.string.message_get_lyrics_failure)
                showSearchResult(r)
            }
        }

        // List item
        searchResultAdapter.itemClickListener = listener@{ _, position ->
            val inputMethodMgr = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            inputMethodMgr.hideSoftInputFromWindow(binding.searchResultListView.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)

            showLyrics(null)

            binding.searchLyricsScrollView.visibility = View.INVISIBLE
            binding.searchLyricsLoadingLayout.visibility = View.VISIBLE

            GlobalScope.launch(Dispatchers.Main) {
                val item = searchResultAdapter.getItem(position)
                item?.lyrics = async(Dispatchers.Default) {
                    return@async ViewLyricsSearcher.downloadLyricsText(item?.lyricURL)
                }.await()
                showLyrics(item)
            }
        }
    }



    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)

        binding.searchResultListView.adapter = searchResultAdapter
        binding.searchTitleEditText.setText(intent.getStringExtra(INTENT_SEARCH_TITLE))
        binding.searchArtistEditText.setText(intent.getStringExtra(INTENT_SEARCH_ARTIST))

        showSearchResult(null)
    }

    /**
     * On create options menu.
     */
    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.activity_search, menu)
        if (menu != null) {
            for (i in 0 until menu.size()) {
                menu.getItem(i).setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
            }
        }
        return true
    }

    /**
     * On options item selected.
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                startActivity(Intent(this, MainActivity::class.java))
                return true
            }
            R.id.menu_search_save_file -> {
                if (!existsLyrics()) {
                    AppUtils.showToast(this, R.string.error_exists_lyrics)
                    return true
                }

                val title = binding.searchTitleEditText.text.toString()
                val artist = binding.searchArtistEditText.text.toString()
                AppUtils.saveFile(this, title, artist)
            }
            R.id.menu_search_save_cache -> {
                if (!existsLyrics()) {
                    AppUtils.showToast(this, R.string.error_exists_lyrics)
                    return true
                }

                val dialog = ConfirmDialogFragment.newInstance(
                        getString(R.string.message_dialog_confirm_save_cache),
                        getString(R.string.label_confirmation),
                        getString(R.string.label_dialog_confirm_save_cache),
                        null,
                        getString(android.R.string.cancel)
                )
                dialog.clickListener = { _, which, _ ->
                    if (which == DialogInterface.BUTTON_POSITIVE) {
                        val title = binding.searchTitleEditText.text.toString()
                        val artist = binding.searchArtistEditText.text.toString()

                        GlobalScope.launch(Dispatchers.Main) {
                            val result = async(Dispatchers.Default) {
                                val cache = SearchCache.create(AppUtils.coalesce(title), AppUtils.coalesce(artist), searchResultAdapter.selectedItem)
                                return@async dao.create(cache) > 0
                            }
                            if (result.await())
                                AppUtils.showToast(this@SearchActivity, R.string.message_save_cache)
                        }
                    }
                }
                dialog.show(this)
            }
            R.id.menu_search_open_cache -> {
                val intent = Intent(this, CacheActivity::class.java)
                intent.putExtra(CacheActivity.INTENT_SEARCH_TITLE, binding.searchTitleEditText.text.toString())
                intent.putExtra(CacheActivity.INTENT_SEARCH_ARTIST, binding.searchArtistEditText.text.toString())
                startActivity(intent)
                return true
            }
        }

        return super.onOptionsItemSelected(item)
    }

    /**
     * On activity result
     */
    public override fun onActivityResult(requestCode: Int, resultCode: Int, resultData: Intent?) {
        if (requestCode == AppUtils.REQUEST_CODE_SAVE_FILE && resultCode == RESULT_OK) {
            // Save to lyrics file
            if (!existsLyrics()) {
                AppUtils.showToast(this, R.string.error_exists_lyrics)
                return
            }

            try {
                val uri = resultData?.data!!
                contentResolver.openOutputStream(uri)!!.bufferedWriter(Charsets.UTF_8).use {
                    it.write(searchResultAdapter.selectedItem?.lyrics)
                }
                AppUtils.showToast(this, R.string.message_lyrics_save_succeeded)
            } catch (e: Exception) {
                Timber.e(e)
                AppUtils.showToast(this, R.string.message_lyrics_save_failed)
            }
        }

        // Hide keyboard
        if (currentFocus != null) {
            val inputMethodMgr = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            inputMethodMgr.hideSoftInputFromWindow(currentFocus?.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)
        }
    }

    /**
     * Show search result list.
     */
    private fun showSearchResult(result: Result?) {
        try {
            searchResultAdapter.clear()
            if (result != null)
                searchResultAdapter.addAll(result.infoList)
            searchResultAdapter.notifyDataSetChanged()
            showLyrics(null)
        } finally {
            binding.searchResultListView.visibility = View.VISIBLE
            binding.searchResultLoadingLayout.visibility = View.INVISIBLE
        }
    }

    /**
     * Show lyrics.
     */
    private fun showLyrics(item: ResultItem?) {
        if (item == null) {
            binding.searchLyricsTextView.text = null
        } else {
            binding.searchLyricsTextView.text = item.lyrics
        }
        searchResultAdapter.selectedItem = item
        if (item != null) // for ripple effect
            searchResultAdapter.notifyDataSetChanged()
        binding.searchLyricsScrollView.visibility = View.VISIBLE
        binding.searchLyricsLoadingLayout.visibility = View.INVISIBLE
    }

    /**
     * Check existence of lyrics
     * @return true if exists lyrics.
     */
    private fun existsLyrics(): Boolean {
        return !searchResultAdapter.selectedItem?.lyricURL.isNullOrEmpty()
    }



    /**
     * Unsent list adapter
     */
    inner class SearchResultAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
        /** Item list */
        private val itemList: MutableList<ResultItem> = mutableListOf()
        /** Selected item.  */
        var selectedItem: ResultItem? = null

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
            val itemBinding: LayoutSearchItemBinding = DataBindingUtil.inflate(LayoutInflater.from(parent.context), R.layout.layout_search_item, parent,false)
            val rootView = itemBinding.root
            rootView.tag = itemBinding
            return object : RecyclerView.ViewHolder(rootView) {}
        }

        @SuppressLint("ClickableViewAccessibility")
        override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
            val listPosition = holder.adapterPosition
            val item = itemList[listPosition]
            (holder.itemView.tag as LayoutSearchItemBinding).apply {
                val context = root.context
                searchItemRadioButton.isChecked = (item == selectedItem)
                searchItemTitleTextView.text = item.musicTitle
                searchItemArtistTextView.text = AppUtils.coalesce(item.musicArtist, "-")
                searchItemAlbumTextView.text = AppUtils.coalesce(item.musicAlbum, "-")
                searchItemDownloadTextView.text = context.getString(R.string.label_search_item_download, item.lyricDownloadsCount)
                searchItemRatingTextView.text = context.getString(R.string.label_search_item_rating, item.lyricRate, item.lyricRatesCount)
                searchItemFromTextView.text = context.getString(R.string.label_search_item_from, item.lyricUploader)

                searchItemRadioButton.setOnTouchListener { _, event ->
                    root.onTouchEvent(event)
                }
                root.setOnClickListener {
                    itemClickListener?.invoke(root, listPosition)
                }
            }
        }

        override fun getItemCount(): Int {
            return itemList.size
        }

        fun getItem(position: Int): ResultItem? {
            return itemList.getOrNull(position)
        }

        fun addAll(list: Collection<ResultItem>?) {
            if (list != null)
                itemList.addAll(list)
        }

        fun clear() {
            itemList.clear()
        }

        var itemClickListener: ((view: View, which: Int) -> Unit)? = null

    }



    companion object {
        const val INTENT_SEARCH_TITLE = "INTENT_SEARCH_TITLE"
        const val INTENT_SEARCH_ARTIST = "INTENT_SEARCH_ARTIST"
    }

}

